<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обработка пропусков</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Замена типа данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обработка дубликатов</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Лемматизация</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Категоризация данных</a></span><ul class="toc-item"><li><span><a href="#Признак-наличия-детей" data-toc-modified-id="Признак-наличия-детей-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>Признак наличия детей</a></span></li><li><span><a href="#Cемейное-положение" data-toc-modified-id="Cемейное-положение-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>Cемейное положение</a></span></li><li><span><a href="#Уровень-дохода" data-toc-modified-id="Уровень-дохода-2.5.3"><span class="toc-item-num">2.5.3&nbsp;&nbsp;</span>Уровень дохода</a></span></li><li><span><a href="#Цель-кредита" data-toc-modified-id="Цель-кредита-2.5.4"><span class="toc-item-num">2.5.4&nbsp;&nbsp;</span>Цель кредита</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.5.5"><span class="toc-item-num">2.5.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></li><li><span><a href="#Ответы-на-вопросы-исследования" data-toc-modified-id="Ответы-на-вопросы-исследования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на вопросы исследования</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Сводные-таблицы-(кредиты-с-задолженностью-по-категориям-заемщиков)" data-toc-modified-id="Сводные-таблицы-(кредиты-с-задолженностью-по-категориям-заемщиков)-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Сводные таблицы (кредиты с задолженностью по категориям заемщиков)</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Загрузка данных

In [1]:
import numpy as np
import pandas as pd

In [2]:
df_borrower = pd.read_csv('datasets/borrower.csv')
df_borrower.info()
df_borrower

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   Среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
0      253875.639453                 покупка жилья  
1      112080.014102       приобретение автомобиля  
2      145885.952297                 покупка жилья  
3      267628.550329    дополнительное образование  
4      158616.077870               сыграть свадьбу  
...              ...                           ...  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля  

[21525 rows x 12 columns]

### Вывод

В таблице 21525 записей о потенциальных заемщиках. 

В таблице 2 вещественных атрибута, 5 целочисленных и 5 строковых.

Чуть более 2000 записей не содержат данных о стаже и доходе. 

## Предобработка данных

### Обработка пропусков

In [3]:
df_borrower[df_borrower['days_employed'].isnull() & df_borrower['total_income'].isnull()]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

In [4]:
df_borrower['total_income'].describe()

count    1.935100e+04
mean     1.674223e+05
std      1.029716e+05
min      2.066726e+04
25%      1.030532e+05
50%      1.450179e+05
75%      2.034351e+05
max      2.265604e+06
Name: total_income, dtype: float64

Пропуски присутсвтуют одновременно в атрибутах days_employed (**общий трудовой стаж в днях**) и total_income (**ежемесячный доход**). Учитывая то, что это данные о действующих клиентах кредитных продуктов банка, скорее всего пропуски в этих записях ошибочны (некорректно заведены анкеты клиентов, технические сбои при выгрузке данных, специально скрыты). 

Вместе с тем, задача исследования сформулирована следующим образом: "влияет ли **семейное положение** и **количество детей** клиента на факт погашения кредита в срок?". В рамках решения этой задачи параметром общего стажа можно пренебречь, соответственно заполнив пропуски нулями. 

In [5]:
df_borrower['days_employed'] = df_borrower['days_employed'].fillna(0)
df_borrower[df_borrower['days_employed'].isnull()]['days_employed'].count()

0

Один из вопросов исследования сформулирован так: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?". Наличие пропусков в значениях ежемесячного дохода может быть существенным для ответа на этот вопрос. 
Определим медианные значения дохода для групп клиентов по параметрам **тип занятости** и **пол** с целью последующей замены пропусков в значениях уровня дохода.

In [6]:
df_borrower[df_borrower['total_income'].notnull()].groupby(['income_type', 'gender'])['total_income'].median()

income_type      gender
безработный      F         202722.511368
                 M          59956.991984
в декрете        F          53829.130729
госслужащий      F         136982.489425
                 M         185964.946748
компаньон        F         160820.776207
                 M         196818.800823
                 XNA       203905.157261
пенсионер        F         115807.789733
                 M         130739.759955
предприниматель  F         499163.144947
сотрудник        F         130615.610597
                 M         162161.177474
студент          M          98201.625314
Name: total_income, dtype: float64

Таблица медианного дохода по группам клиентов показала наличие артефакта в атрибуте 'gender' - значения 'XNA'. Посмотрим на записи с этим значением:

In [7]:
df_borrower[df_borrower['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10701         0   -2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

Такая запись всего одна. Уровень дохода для этого клиента чуть выше медианного для группы (компаньон, мужчина), поэтому заменим указанное значение пола на M.

In [8]:
df_borrower.loc[df_borrower['gender'] == 'XNA', 'gender'] = 'M'
df_borrower[df_borrower['gender'] == 'XNA']['gender'].count()

0

Артефакт устранен. Снова пересчитаем медианное значение дохода для групп клиентов.

In [9]:
df_median = df_borrower[df_borrower['total_income'].notnull()].groupby(['income_type', 'gender'])['total_income'].median().reset_index()
df_median

income_type gender   total_income
0       безработный      F  202722.511368
1       безработный      M   59956.991984
2         в декрете      F   53829.130729
3       госслужащий      F  136982.489425
4       госслужащий      M  185964.946748
5         компаньон      F  160820.776207
6         компаньон      M  196864.845426
7         пенсионер      F  115807.789733
8         пенсионер      M  130739.759955
9   предприниматель      F  499163.144947
10        сотрудник      F  130615.610597
11        сотрудник      M  162161.177474
12          студент      M   98201.625314

In [10]:
# функция возвращает медианное значение дохода для входных параметров 
# incom_type - тип занятости
# gender - пол
# при отсутствии в таблице медианных значений указанной группы - возврат медианного значения дохода 
# по всей таблице (без учета пустых значений)

def find_median_value(incom_type, gender)->float:
    median_group = df_median[(df_median['income_type'] == incom_type) & (df_median['gender'] == gender)]['total_income']
    if median_group.empty:
        return df_borrower[df_borrower['total_income'].notnull()]['total_income'].median()
    else:
        return median_group.values[0]
    

In [11]:
df_borrower.loc[df_borrower['total_income'].isnull(), 'total_income'] = df_borrower.apply(
    lambda row: find_median_value(row['income_type'], row['gender']),
    axis=1
)
df_borrower[df_borrower['total_income'].isnull()]['total_income'].count()


0

<font color='green'>Классная  реализация. Так пропуски будут заполнены максимально достоверно.

In [12]:
df_borrower[df_borrower['days_employed'].isnull() & df_borrower['total_income'].isnull()]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

Проверка - пропусков в таблице не осталось. Визуально посмотрим на ту часть таблицы, где раньше были пропуски (сравним значение в атрибуте total_income и таблице df_median для соответствующей группы):

In [13]:
df_borrower[df_borrower['days_employed'] == 0].head(5)

children  days_employed  dob_years education  education_id  \
12         0            0.0         65   среднее             1   
26         0            0.0         41   среднее             1   
29         0            0.0         63   среднее             1   
41         0            0.0         50   среднее             1   
55         0            0.0         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

     total_income                           purpose  
12  130739.759955                   сыграть свадьбу  
26  185964.946748                       образование  
29  115807.789733  строительство жилой недвижимости  
41  136982.489425  сделка с подержанным автомобилем  
55  115807.789733                   сыграть свадьбу

#### Вывод

Пропуски в таблице были одновременно в 2 атрибутах (общий стаж и уровень ежемесячного дохода) примерно в 10% записей. 

Атрибут общего стажа является незначимым в рамках исследования, соответственно пропуски в значениях стажа заменены нулями.

Исследование зависимости между уровнем дохода и возвратом кредита в срок является одной из задач исследования. Доля данных с пустыми значениями составляет примерно 10% от общего объема, что не позволяет выполнить замену пропусков нулями. Для замены пустот в уровне дохода проведена группировка клиентов по атрибутам (пол, тип занятости) и определены медианные значения для каждой группы. Указанные медианные значения вставлены вместо пустот в соответствующих записях таблицы.

### Замена типа данных

In [14]:
df_borrower.info()
df_borrower.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

В таблице присутсвуют два вещественных атрибута: days_employed (**общий трудовой стаж в днях**) и total_income (**ежемесячный доход**)

Поскольку общий трудовой стаж указан в днях, целесообразно приведение указанного атрибута к целочисленному типу данных.

Поскольку ежемесячный доход указан в рублях (вещественная часть содержит копейки и доли копеек), можно также выполнить его приведение к целочисленному типу данных.

Все другие строковые атрибуты фактически содержат строковые значения, нетребующие преобразования в целочисленный тип.

Преобразование будем проводить методом astype() с параметром 'int'. Возможные потери точности для нас несущественны. 

In [15]:
df_borrower['total_income'] = df_borrower['total_income'].astype('int')
df_borrower['days_employed'] = df_borrower['days_employed'].astype('int')
df_borrower.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int64 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int64 
 11  purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


Преобразование в целочисленный тип проведено успешно. Теперь посмотрим статистику данных в атрибутах стажа и ежемесячного дохода.

In [16]:
df_borrower['days_employed'].describe()

count     21525.000000
mean      56679.163252
std      134870.418456
min      -18388.000000
25%       -2518.000000
50%        -982.000000
75%           0.000000
max      401755.000000
Name: days_employed, dtype: float64

В атрибуте 'days_employed' присутствует артефакт - отрицательные значения. Вероятнее всего, что это ошибка ввода данных. Заменим отрицательные значения на значения модуля числа.

In [17]:
df_borrower['days_employed'] = abs(df_borrower['days_employed'])
df_borrower['days_employed'].describe()

count     21525.000000
mean      60155.970128
std      133355.906156
min           0.000000
25%         610.000000
50%        1808.000000
75%        4779.000000
max      401755.000000
Name: days_employed, dtype: float64

Проверяем - отрицательных значений стажа больше нет. Вместе с тем, в данных присутствуют значения, которые изначально были положительными (900 лет стажа) но также далеки от достоверности. Поскольку в рамках исследования корреляция со стажем не является предметом исследования, оставим эти значения без изменений. В противном случае было бы целесообразно их заменить на некоторое разумное значение, соответствующее максимально возможному стажу.

#### Вывод

Значения атрибутов трудового стажа и ежемесячного дохода переведены в целочисленный тип без существенной в рамках исследования потери точности. Устранен артефакт - наличие отрицательных значений трудового стажа.

### Обработка дубликатов

In [18]:
df_borrower.duplicated().sum()


54

Дубликаты в таблице есть (54 записи). Сделаем предварительную оценку дубликатов.

In [19]:
df_duplicated = df_borrower[df_borrower.duplicated()]
df_duplicated.head(5)

children  days_employed  dob_years education  education_id  \
2849         0              0         41   среднее             1   
4182         1              0         34    ВЫСШЕЕ             0   
4851         0              0         60   среднее             1   
5557         0              0         58   среднее             1   
7808         0              0         57   среднее             1   

         family_status  family_status_id gender income_type  debt  \
2849   женат / замужем                 0      F   сотрудник     0   
4182  гражданский брак                 1      F   сотрудник     0   
4851  гражданский брак                 1      F   пенсионер     0   
5557  гражданский брак                 1      F   пенсионер     0   
7808  гражданский брак                 1      F   пенсионер     0   

      total_income                  purpose  
2849        130615  покупка жилья для семьи  
4182        130615                  свадьба  
4851        115807                  свадьба  
5557        115807          сыграть свадьбу  
7808        115807    на проведение свадьбы

Гипотеза - дубликаты существуют только в тех записях, для которых изначально не были указаны общий стаж и уровень дохода. Проверяем (считаем количество дубликатов в записях, для которых общий стаж больше 0).

In [20]:
df_duplicated[df_duplicated['days_employed'] > 0]['days_employed'].count()

0

Так и есть.

Допускаю, что в исходных данных среди этого количества дубликатов есть физически разные объекты (разные люди одного возраста, с одним и тем же семейным положением, уровнем образования, количеством детей, типом дохода, взявшие кредит на одну и ту же цель). Скорее всего доля таких совпадений все-таки небольшая, но все же попробуем ее оценить по оставшейся части (без исходных null).

In [21]:
df_notnull = df_borrower[df_borrower['days_employed'] > 0]
df_notnull[['children', 'dob_years', 'education_id', 'family_status_id', 'gender', 'income_type', 'debt', 'purpose']].duplicated().sum() / df_notnull.shape[0]


0.15544416309234665

Получаем значение в 15% возможных совпадений в атрибутах таблицы (без учета атрибутов стажа и ежемесячного дохода). Т.е. в нашем случае если мы просто удалим дубликаты (54 записи, определенные выше), мы возможно потеряем 8 записей о действительно разных людях, для которых не были указаны стаж и доход и которые совпали по остальным атрибутам. В общем объеме (20к записей) они составляют примерно 0.05%, ими можно пренебречь. Удаляем дубликаты.

In [22]:
df_borrower = df_borrower.drop_duplicates()
df_borrower.duplicated().sum()

0

Дубликаты удалены.

In [23]:
df_borrower.shape[0]

21471

Выделим словарь по типу образованию и после этого повторно повторим на наличие дубликатов.

In [24]:
df_dict_edu = df_borrower.groupby(['education_id', 'education'])['education_id', 'education'].size().reset_index()
df_dict_edu

<ipython-input-24-183d6f2ca4dd>:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df_dict_edu = df_borrower.groupby(['education_id', 'education'])['education_id', 'education'].size().reset_index()


education_id            education      0
0              0               ВЫСШЕЕ    273
1              0               Высшее    268
2              0               высшее   4710
3              1              СРЕДНЕЕ    772
4              1              Среднее    711
5              1              среднее  13705
6              2  НЕОКОНЧЕННОЕ ВЫСШЕЕ     29
7              2  Неоконченное высшее     47
8              2  неоконченное высшее    668
9              3            НАЧАЛЬНОЕ     17
10             3            Начальное     15
11             3            начальное    250
12             4       УЧЕНАЯ СТЕПЕНЬ      1
13             4       Ученая степень      1
14             4       ученая степень      4

В исходной таблице в атрибуте education встречаются различные написания уровня образования (например, "высшее" и "ВЫСШЕЕ"), вместе с тем идентификатор уровня образования задан верно. Целесообразно вынестри строковое значения уровня образования в словарь, оставив в исходной таблице только идентификаторы.

In [25]:
df_dict_edu['education'] = df_dict_edu.apply(lambda row : str.lower(row['education']), axis=1)
df_dict_edu

education_id            education      0
0              0               высшее    273
1              0               высшее    268
2              0               высшее   4710
3              1              среднее    772
4              1              среднее    711
5              1              среднее  13705
6              2  неоконченное высшее     29
7              2  неоконченное высшее     47
8              2  неоконченное высшее    668
9              3            начальное     17
10             3            начальное     15
11             3            начальное    250
12             4       ученая степень      1
13             4       ученая степень      1
14             4       ученая степень      4

In [26]:
df_dict_edu = df_dict_edu[['education_id', 'education']].drop_duplicates()
df_dict_edu

education_id            education
0              0               высшее
3              1              среднее
6              2  неоконченное высшее
9              3            начальное
12             4       ученая степень

Теперь у нас сформировань словарь вида: идентификатор - название типа образования. В исходной таблице можем удалить атрибут с названием типа образования.

In [27]:
df_borrower = df_borrower[['children', 'days_employed', 'dob_years', 'education_id', 'family_status', \
                           'family_status_id', 'gender', 'income_type', 'debt', 'total_income', 'purpose']]

df_borrower.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21471 entries, 0 to 21524
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21471 non-null  int64 
 1   days_employed     21471 non-null  int64 
 2   dob_years         21471 non-null  int64 
 3   education_id      21471 non-null  int64 
 4   family_status     21471 non-null  object
 5   family_status_id  21471 non-null  int64 
 6   gender            21471 non-null  object
 7   income_type       21471 non-null  object
 8   debt              21471 non-null  int64 
 9   total_income      21471 non-null  int64 
 10  purpose           21471 non-null  object
dtypes: int64(7), object(4)
memory usage: 2.0+ MB


И повторно проведем поиск и удаление дубликатов.

In [28]:
df_borrower.duplicated().sum()

17

In [29]:
df_duplicated = df_borrower[df_borrower.duplicated()]
df_duplicated

children  days_employed  dob_years  education_id     family_status  \
3290          0              0         58             1  гражданский брак   
6312          0              0         30             1   женат / замужем   
7921          0              0         64             0  гражданский брак   
7938          0              0         71             1  гражданский брак   
9604          0              0         71             1  гражданский брак   
9855          0              0         62             1   женат / замужем   
14097         0              0         48             1  гражданский брак   
14728         0              0         46             1  гражданский брак   
15991         0              0         51             1  гражданский брак   
16204         0              0         56             1   женат / замужем   
17338         0              0         64             1  гражданский брак   
19295         0              0         51             1   женат / замужем   
19369         0              0         45             1  гражданский брак   
19559         0              0         41             1   женат / замужем   
20187         0              0         65             1  гражданский брак   
20297         1              0         23             1  гражданский брак   
20662         0              0         58             1  гражданский брак   

       family_status_id gender income_type  debt  total_income  \
3290                  1      F   пенсионер     0        115807   
6312                  0      M   сотрудник     0        162161   
7921                  1      F   пенсионер     0        115807   
7938                  1      F   пенсионер     0        115807   
9604                  1      F   пенсионер     0        115807   
9855                  0      F   пенсионер     0        115807   
14097                 1      F   сотрудник     0        130615   
14728                 1      F   сотрудник     0        130615   
15991                 1      F   компаньон     0        160820   
16204                 0      F   пенсионер     0        115807   
17338                 1      F   пенсионер     0        115807   
19295                 0      F   сотрудник     0        130615   
19369                 1      F   компаньон     0        160820   
19559                 0      F   сотрудник     0        130615   
20187                 1      F   пенсионер     0        115807   
20297                 1      F   сотрудник     0        130615   
20662                 1      M   сотрудник     0        162161   

                                     purpose  
3290                         сыграть свадьбу  
6312        строительство жилой недвижимости  
7921                   на проведение свадьбы  
7938                   на проведение свадьбы  
9604                   на проведение свадьбы  
9855   получение дополнительного образования  
14097                                свадьба  
14728                покупка жилья для сдачи  
15991                  на проведение свадьбы  
16204                  на покупку автомобиля  
17338                        сыграть свадьбу  
19295      покупка коммерческой недвижимости  
19369                                свадьба  
19559                        свой автомобиль  
20187                        сыграть свадьбу  
20297                        сыграть свадьбу  
20662                  на проведение свадьбы

Повторно удалим дубликаты, появившиеся вследствие удаления атрибута с названием типа образования.

In [30]:
df_borrower = df_borrower.drop_duplicates()
df_borrower.duplicated().sum()

0

In [31]:
df_borrower.shape[0]

21454

#### Вывод

В таблице обнаружен 71 дубликат, все они - среди записей, для которых изначально в атрибутах общего стажа и уровня дохода были пропуски. 

Проведена оценка количества значимых записей, которые будут удалены при удалении дубликатов - их доля составила 0,05% от общего количества записей (пренебрежимо малое количество). Выполнено удаление дубликатов. В таблице осталась 21454 запись.

### Лемматизация

Определим, на какие основные цели брали кредиты клиенты банка. 

In [32]:
from pymystem3 import Mystem
from collections import Counter
mys = Mystem() 

In [33]:
purpose = df_borrower['purpose'].unique()
purpose

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Сначала определим все уникальные цели кредита, указанные в атрибуте 'purpose', и проанализируем их.

Очевидно, что в таблице много целей, которые хотелось бы отнести к одной категории, например:
- к категории 'автомобиль': 'приобретение автомобиля', 'автомобили', 'сделка с подержанным автомобилем' и т.д.
- к категории 'недвижимость': 'операции с недвижимостью', 'строительство собственной недвижимости' и т.д.

Для решения этой задачи воспользуемся алгоритмом лемматизации текста. Отсортируем по частотам встречаемости все леммы, выделенные в указанных выше целях:

In [34]:
lemmas = mys.lemmatize(' '.join(purpose))
Counter(lemmas).most_common()

[(' ', 96),
 ('покупка', 10),
 ('недвижимость', 10),
 ('автомобиль', 9),
 ('образование', 9),
 ('жилье', 7),
 ('с', 5),
 ('операция', 4),
 ('на', 4),
 ('свой', 4),
 ('свадьба', 3),
 ('строительство', 3),
 ('получение', 3),
 ('высокий', 3),
 ('дополнительный', 2),
 ('для', 2),
 ('коммерческий', 2),
 ('жилой', 2),
 ('подержать', 2),
 ('заниматься', 2),
 ('сделка', 2),
 ('приобретение', 1),
 ('сыграть', 1),
 ('проведение', 1),
 ('семья', 1),
 ('собственный', 1),
 ('со', 1),
 ('профильный', 1),
 ('сдача', 1),
 ('ремонт', 1),
 ('\n', 1)]

Проанализируем результат. Среди указанных значений можно определить 4 больших категорий целей кредита:
- 'недвижимость'
- 'автомобиль'
- 'образование'
- 'свадьба'

Определим функцию, которая по входной строке будет определять категорию кредита (из указанных выше).

In [35]:
# функция возвращает название категории цели кредита по строке (цель в произвольной форме) 
# purpose_text - цель кредита в произвольной форме

def find_purpose(purpose_text):
    purpose_lemm = mys.lemmatize(purpose_text)
    if ('недвижимость' in purpose_lemm) | ('жилье' in purpose_lemm):
        return 'недвижимость'
    elif 'автомобиль' in purpose_lemm:
        return 'автомобиль'
    elif 'образование' in purpose_lemm:
        return 'образование'   
    elif 'свадьба' in purpose_lemm:
        return 'свадьба'       

Создадим таблицу с атрибутами **цели кредита** (purpose_text) и выделенной **категории кредита** (purpose)

In [36]:
df_purpose = pd.DataFrame(data=purpose, columns={'purpose_text'})
df_purpose['purpose_category'] = df_purpose.apply(
    lambda row: find_purpose(row['purpose_text']),
    axis=1
)

df_purpose

purpose_text purpose_category
0                            покупка жилья     недвижимость
1                  приобретение автомобиля       автомобиль
2               дополнительное образование      образование
3                          сыграть свадьбу          свадьба
4                        операции с жильем     недвижимость
5                              образование      образование
6                    на проведение свадьбы          свадьба
7                  покупка жилья для семьи     недвижимость
8                     покупка недвижимости     недвижимость
9        покупка коммерческой недвижимости     недвижимость
10              покупка жилой недвижимости     недвижимость
11  строительство собственной недвижимости     недвижимость
12                            недвижимость     недвижимость
13              строительство недвижимости     недвижимость
14      на покупку подержанного автомобиля       автомобиль
15            на покупку своего автомобиля       автомобиль
16   операции с коммерческой недвижимостью     недвижимость
17        строительство жилой недвижимости     недвижимость
18                                   жилье     недвижимость
19         операции со своей недвижимостью     недвижимость
20                              автомобили       автомобиль
21                   заняться образованием      образование
22        сделка с подержанным автомобилем       автомобиль
23                   получение образования      образование
24                              автомобиль       автомобиль
25                                 свадьба          свадьба
26   получение дополнительного образования      образование
27                    покупка своего жилья     недвижимость
28                операции с недвижимостью     недвижимость
29           получение высшего образования      образование
30                         свой автомобиль       автомобиль
31                    сделка с автомобилем       автомобиль
32                  профильное образование      образование
33                      высшее образование      образование
34                 покупка жилья для сдачи     недвижимость
35                   на покупку автомобиля       автомобиль
36                            ремонт жилью     недвижимость
37            заняться высшим образованием      образование

Визуально проверяем - все цели правильно отнесены к соответствующей категории.

Добавим выделенные категории кредитов к исходной таблице.

In [37]:
df_result = df_borrower.merge(df_purpose, how='inner', left_on='purpose', right_on='purpose_text')
df_result.head(5)

children  days_employed  dob_years  education_id     family_status  \
0         1           8437         42             0   женат / замужем   
1         0           5623         33             1   женат / замужем   
2         0            926         27             0  гражданский брак   
3         0           1548         48             1   женат / замужем   
4         0            413         41             1   женат / замужем   

   family_status_id gender  income_type  debt  total_income        purpose  \
0                 0      F    сотрудник     0        253875  покупка жилья   
1                 0      M    сотрудник     0        145885  покупка жилья   
2                 1      M    компаньон     0        255763  покупка жилья   
3                 0      F    компаньон     0        157245  покупка жилья   
4                 0      M  госслужащий     0        118551  покупка жилья   

    purpose_text purpose_category  
0  покупка жилья     недвижимость  
1  покупка жилья     недвижимость  
2  покупка жилья     недвижимость  
3  покупка жилья     недвижимость  
4  покупка жилья     недвижимость

In [38]:
df_result.shape[0]

21454

In [39]:
df_result['purpose_category'].value_counts()

недвижимость    10811
автомобиль       4306
образование      4013
свадьба          2324
Name: purpose_category, dtype: int64

Наиболее востребованными у клиентов оказались кредиты на недвижимость, далее - автокредиты, кредиты на образование и кредиты на свадьбу. Доля кредитов на свадьбу достаточно существенна (~10%), чтобы ее можно было пренебречь (или отнести к другой категории).

#### Вывод

На основе целей кредита, указанных в таблице, путем лемматизации определены 4 основные категории кредита.

Проведено сопоставление целей и выделенных категорий. Атрибут категории кредита добавлен в общую таблицу с информацией о клиентах.

### Категоризация данных

#### Признак наличия детей

Категоризацию данных целесообразно проводить в разрере вопросов, сформулированных в рамках исследования. 

**Для ответа на вопрос №1 нам понадобится атрибут - признак наличия детей**. 

Сформируем его на основе атрибута количества детей.

In [40]:
df_result['children'].value_counts()

 0     14091
 1      4808
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Сначала проверим множество значений атрибута **количество детей**. Устраним артефакты (значение -1 заменим на 1, 20 - на 2).

In [41]:
df_result.loc[df_result['children'] == -1, 'children'] = 1
df_result.loc[df_result['children'] == 20, 'children'] = 2
df_result['children'].value_counts()

0    14091
1     4855
2     2128
3      330
4       41
5        9
Name: children, dtype: int64

<font color='purple'>Определим три категории клиентов по наличию (количеству) детей: семьи без детей, с 1-2 детьми и многодетные.

In [42]:
# функция возвращает название категории по количеству детей
# purpose_text - цель кредита в произвольной форме

def get_children_count_level(children_count):
    if children_count == 0:
        return 'нет детей'
    elif children_count <= 2:
        return '1-2 ребенка'
    else:
        return 'многодетные'

In [43]:
df_result['has_children'] = df_result.apply(
    lambda row: get_children_count_level(row['children']),
    axis=1
)

df_result['has_children'].value_counts()

нет детей      14091
1-2 ребенка     6983
многодетные      380
Name: has_children, dtype: int64

Атрибут категории по наличию детей заполнен.

#### Cемейное положение

**Для ответа на вопрос №2 нам понадобится атрибут - семейное положение**. 

Проверим значения в атрибуах семейного положения. 

In [44]:
df_result.groupby(['family_status_id', 'family_status'])['family_status_id', 'family_status'].count()

<ipython-input-44-1b6a29eee5cd>:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df_result.groupby(['family_status_id', 'family_status'])['family_status_id', 'family_status'].count()


family_status_id  family_status
family_status_id family_status                                         
0                женат / замужем                   12339          12339
1                гражданский брак                   4151           4151
2                вдовец / вдова                      959            959
3                в разводе                          1195           1195
4                Не женат / не замужем              2810           2810

Артефактов в значениях указанных атрибутов нет. <font color='purple'>Выделим семейные статусы в отдельный словарь и удалим названия семейного статуса из исходной таблицы.

In [45]:
df_dict_family = df_borrower.groupby(['family_status_id', 'family_status'])['family_status_id', 'family_status'].size().reset_index()[['family_status_id', 'family_status']]
df_dict_family

<ipython-input-45-fd3e7df1625c>:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df_dict_family = df_borrower.groupby(['family_status_id', 'family_status'])['family_status_id', 'family_status'].size().reset_index()[['family_status_id', 'family_status']]


family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

In [46]:
df_borrower = df_borrower[['children', 'days_employed', 'dob_years', 'education_id', \
                           'family_status_id', 'gender', 'income_type', 'debt', 'total_income', 'purpose']]

df_borrower.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21454 entries, 0 to 21524
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21454 non-null  int64 
 1   days_employed     21454 non-null  int64 
 2   dob_years         21454 non-null  int64 
 3   education_id      21454 non-null  int64 
 4   family_status_id  21454 non-null  int64 
 5   gender            21454 non-null  object
 6   income_type       21454 non-null  object
 7   debt              21454 non-null  int64 
 8   total_income      21454 non-null  int64 
 9   purpose           21454 non-null  object
dtypes: int64(7), object(3)
memory usage: 1.8+ MB


В рамках исследования рассмотрим два способа деления клиентов на категории по семейному положению:
- способ 1: "женат / замужем (в т.ч. гражданский брак)" и "не женат / не замужем (в т.ч. вдовцы, в разводе)"
- способ 2: "когда-либо состоял / состоит в официальных отношених (в т.ч. вдовцы, в разводе)" и "никогда не состоял / не состоит в официальных отношених (в т.ч. гражданский брак)"

In [47]:
df_result['married_now'] = df_result.apply(
    lambda row: 'семейный' if row['family_status_id'] in [0, 1] else 'холостой',
    axis=1
)

df_result['married_now'].value_counts()

семейный    16490
холостой     4964
Name: married_now, dtype: int64

In [48]:
df_result['married_ever'] = df_result.apply(
    lambda row: 'когда-либо был в браке' if row['family_status_id'] in [0, 2, 3] else 'никогда не был в браке',
    axis=1
)

df_result['married_ever'].value_counts()

когда-либо был в браке    14493
никогда не был в браке     6961
Name: married_ever, dtype: int64

Атрибут категории по семейному положению заполнен.

#### Уровень дохода 

**Для ответа на вопрос №3 нам понадобится атрибут - уровень дохода**. 

Проверим значения в атрибуах уровня дохода. 

In [49]:
df_result['total_income'].describe()

count    2.145400e+04
mean     1.653161e+05
std      9.828187e+04
min      2.066700e+04
25%      1.076230e+05
50%      1.441265e+05
75%      1.968640e+05
max      2.265604e+06
Name: total_income, dtype: float64

In [50]:
df_result['total_income'].median()

144126.5

Минимальным значением уровня дохода является значение ~20.000, максимальным - ~2.265.000, медианным - ~144.000.

Можно воспользоваться медианным значением и разделить всех клиентов на две группы: с низким и высоким уровнем дохода. Но наверно такое деление будет не совсем показательным.

Выделим 4 группы по уровню дохода: 
- первые 20% (низкий доход)
- группа 65% (диапазон 20%-85%) в середине (средний доход)
- группа 13% (диапазон 85%-98%) в середине (высокий доход)   
- и 2% с наиболее высоким уровнем дохода (сверхвысокий доход)

In [51]:
quantile_20 = df_result['total_income'].quantile(.2) 
quantile_20

98537.6

In [52]:
quantile_85 = df_result['total_income'].quantile(.85) 
quantile_85

237537.1

In [53]:
quantile_98 = df_result['total_income'].quantile(.98) 
quantile_98

421410.71999999974

In [54]:
# функция возвращает название категории уровня дохода по значению дохода
# purpose_text - цель кредита в произвольной форме

def get_income_level(total_income):
    if total_income <= quantile_20:
        return 'низкий'
    elif total_income <= quantile_85:
        return 'средний'
    elif total_income <= quantile_98:
        return 'высокий'
    else:
        return 'сверхвысокий'

In [55]:
get_income_level(1000000)

'сверхвысокий'

In [56]:
df_result['income_level'] = df_result.apply(
    lambda row: get_income_level(row['total_income']),
    axis=1
)

In [57]:
df_result['income_level'].value_counts()

средний         13945
низкий           4291
высокий          2788
сверхвысокий      430
Name: income_level, dtype: int64

Атрибут категории по уровню дохода заполнен.

#### Цель кредита

**Для ответа на вопрос №4 нам понадобится атрибут - цель кредита**. 

Указанный атрибут сформирован выше, при лемматизации строковых целей кредита. 

#### Вывод

Для ответа на вопросы исследования в таблице определены и заполнены значениями следующие категории данных:
- по количеству детей: нет детей / 1-2 ребенка в семье / многодетные
- по семейному положению (в настоящий момент времени): семейные / холостые
- по семейному положению (когда-либо): состояли (состоят) в браке / не состояли в браке 
- по уровню дохода: низкий / средний / высокий / сверхвысокий
- по целям кредита: недвижимость / автомобиль / образование / свадьба

## Ответы на вопросы исследования

### Есть ли зависимость между наличием детей и возвратом кредита в срок?

Определяем долю тех, кто имел задолженности по кредитам, в зависимости от наличия детей.

In [58]:
df_result[df_result['debt'] == 1].groupby(['has_children'])['has_children'].count() / df_result.groupby(['has_children'])['has_children'].count()

has_children
1-2 ребенка    0.092654
многодетные    0.081579
нет детей      0.075438
Name: has_children, dtype: float64

Доля клиентов с 3 и более детьми, имевших задолженности по кредиту, составляет 8.2%

Доля клиентов с 1-2 детьми, имевших задолженности по кредиту, составляет 9.3%

Доля клиентов без детей, имевших задолженности по кредиту, составляет 7.5%

Вывод: Группа клиентов с 1-2 детьми чаще имеют задолженности по кредитам как относительно клиентов без детей, так и относительно многодетных клиентов. Относительно невысокий процент задолженности в группе многодетных клиентов может объясняться тем, что многодетные клиенты чаще берут кредиты на меньшую сумму, но эта гипотеза требует дальнейших исследований (в нашем случае не хватает данных о суммах кредитов).

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

Определяем долю тех, кто имел задолженности по кредитам, в зависимости от <font color='purple'>своего текущего</font> семейного положения.

In [59]:
df_result[df_result['debt'] == 1].groupby(['married_now'])['married_now'].count() / df_result.groupby(['married_now'])['married_now'].count()

married_now
семейный    0.079988
холостой    0.085012
Name: married_now, dtype: float64

Определяем долю тех клиентов, кто имел задолженности по кредитам, в зависимости от того, состоял ли клиент когда-либо (в т.ч. в настояещее время) в официальных отношениях.

In [60]:
df_result[df_result['debt'] == 1].groupby(['married_ever'])['married_ever'].count() / df_result.groupby(['married_ever'])['married_ever'].count()

married_ever
когда-либо был в браке    0.074450
никогда не был в браке    0.095101
Name: married_ever, dtype: float64

А также посмотрим процент задолженности по исходным семейным статусам.

In [61]:
df_result_family = df_result[df_result['debt'] == 1].groupby(['family_status_id'])['family_status_id'].count() / df_result.groupby(['family_status_id'])['family_status_id'].count().rename("percent")
df_dict_family.merge(df_result_family.to_frame().reset_index(), how='inner',  left_on='family_status_id', right_on='family_status_id')

family_status_id          family_status         0
0                 0        женат / замужем  0.075452
1                 1       гражданский брак  0.093471
2                 2         вдовец / вдова  0.065693
3                 3              в разводе  0.071130
4                 4  Не женат / не замужем  0.097509

По первому типу выделения категорий:

Доля семейных клиентов, имевших задолженности по кредиту, составляет 8.0%

Доля холостых клиентов, имевших задолженности по кредиту, составляет 8.5%
    
По второму типу выделения категорий:

Доля клиентов, когда-либо состоявших в официальных отношениях и имевших задолженности по кредиту, составляет 7.4%

Доля клиентов, никогда не состоявших в официальных отношениях и имевших задолженности по кредиту, составляет 9.5%

Вывод: Холостые клиенты чаще имеют задолженности по кредитам относительно семейных клиентов. Клиенты, никогда не состоявшие в официальных отношениях, имеют самый высокий процент задолженности. 
С учетом значений задолженности по семейному статусу, приведенному в исходной таблице, и в двух выделенных категориях можно следующим образом отранжировать клиентов по увеличению процента задолженности:
- клиенты, состоявшие ранее в официальном браке, но не состоящие в текущий момент (разведены / вдовцы)
- клиенты, состоящие в официальном браке в настоящий момент времени
- клиенты, никогда не состоявшие в официальных отношениях

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Определяем долю тех, кто имел задолженности по кредитам, в зависимости от уровня дохода.

In [62]:
df_result[df_result['debt'] == 1].groupby(['income_level'])['income_level'].count() / df_result.groupby(['income_level'])['income_level'].count()

income_level
высокий         0.072453
низкий          0.080168
сверхвысокий    0.055814
средний         0.083973
Name: income_level, dtype: float64

Доля клиентов с низким уровнем дохода, имевших задолженности по кредиту, составляет 8.0%

Доля клиентов со средним уровнем дохода, имевших задолженности по кредиту, составляет 8.4%

Доля клиентов с высоким уровнем дохода, имевших задолженности по кредиту, составляет 7.2%

Доля клиентов со сверхвысоким уровнем дохода, имевших задолженности по кредиту, составляет 5.6%

Вывод: Клиенты с высоким уровнем дохода реже имеют задолженности по кредитам относительно клиентов с низким и средним уровнем дохода. Вместе с тем, доля клиентов с задолженностями в среднем сегменте немного выше доли клиентов с низким уровнем дохода. Можно предположить, что при другой границе разбиения низкий/средний доход доля клиентов с задолженностями в этих двух сегментах будет одинаковой. 
Одновременно следует отметить, что клиенты со сверхвысоким уровнем дохода имеют наименьший процент задолженности по кредитам. 

### Как разные цели кредита влияют на его возврат в срок?

Определяем долю тех, кто имел задолженности по кредитам, в зависимости от цели кредита.

In [63]:
df_result[df_result['debt'] == 1].groupby(['purpose_category'])['purpose_category'].count() / df_result.groupby(['purpose_category'])['purpose_category'].count()

purpose_category
автомобиль      0.093590
недвижимость    0.072334
образование     0.092200
свадьба         0.080034
Name: purpose_category, dtype: float64

Наибольшее количество задолженностей для кредитов в категориях "автокредит" и "образование" (9.4% и 9.2% соответственно).

Доля задолженностей в категории свадебных кредитов составляет 8.0%

Наименьшее количество задолженностей в категории кредитов на недвижимость - 7.2%

Вывод: Клиенты более ответственно относятся к погашению кредитов на недвижимость (вероятно это связано с тем, что в основном объектом залога является объект недвижимости, в котором клиент проживает - соответственно риск неуплаты для клиента связан с возможной потерей жилья) и свадебных кредит (это вероятно связано с размером выдаваемого кредита - в представленных данных таких сведений нет, но в целом обычно траты на свадьбу существенно ниже затрат на покупку недвижимости или автомобиля).

### Сводные таблицы (кредиты с задолженностью по категориям заемщиков)

При построении сводных таблиц воспользуемся следующим свойством: поскольку в атрибуте 'debt' значение '1' соответствует кредиту с задолженностью, то арифметическое среднее этого атрибута для группы будет соответствовать отношению кредитов с задолженностью к общему числу кредитов (сумма единиц, разделенная на общее количество записей).

Сводная таблица кредитов с задолженностью в зависимости от количества детей:

In [64]:
pd.pivot_table(df_result, values='debt', index=['has_children'], aggfunc={'debt':[np.average]}).reset_index() \
                   .rename(columns={"has_children": "Количество детей", "average": "Доля кредитов с задолженностью"}) \
                   .sort_values(by='Доля кредитов с задолженностью')

Количество детей  Доля кредитов с задолженностью
2        нет детей                        0.075438
1      многодетные                        0.081579
0      1-2 ребенка                        0.092654

Сводная таблица кредитов с задолженностью по отношению к семейному положению клиента в настоящий момент времени:

In [65]:
pd.pivot_table(df_result, values='debt', index=['married_now'], aggfunc={'debt':[np.average]}).reset_index() \
                   .rename(columns={"married_now": "В настоящий момент", "average": "Доля кредитов с задолженностью"}) \
                   .sort_values(by='Доля кредитов с задолженностью')

В настоящий момент  Доля кредитов с задолженностью
0           семейный                        0.079988
1           холостой                        0.085012

Сводная таблица кредитов с задолженностью по отношению к семейному положению клиента в течение жизни:

In [66]:
pd.pivot_table(df_result, values='debt', index=['married_ever'], aggfunc={'debt':[np.average]}).reset_index() \
                   .rename(columns={"married_ever": "В настоящий момент", "average": "Доля кредитов с задолженностью"}) \
                   .sort_values(by='Доля кредитов с задолженностью')

В настоящий момент  Доля кредитов с задолженностью
0  когда-либо был в браке                        0.074450
1  никогда не был в браке                        0.095101

Сводная таблица кредитов с задолженностью в зависимости от уровня дохода:

In [67]:
pd.pivot_table(df_result, values='debt', index=['income_level'], aggfunc={'debt':[np.average]}).reset_index() \
                   .rename(columns={"income_level": "Уровень дохода", "average": "Доля кредитов с задолженностью"}) \
                   .sort_values(by='Доля кредитов с задолженностью')

Уровень дохода  Доля кредитов с задолженностью
2   сверхвысокий                        0.055814
0        высокий                        0.072453
1         низкий                        0.080168
3        средний                        0.083973

Сводная таблица кредитов с задолженностью в зависимости от цели кредита:

In [68]:
pd.pivot_table(df_result, values='debt', index=['purpose_category'], aggfunc={'debt':[np.average]}).reset_index() \
                   .rename(columns={"purpose_category": "Цель кредита", "average": "Доля кредитов с задолженностью"}) \
                   .sort_values(by='Доля кредитов с задолженностью')

Цель кредита  Доля кредитов с задолженностью
1  недвижимость                        0.072334
3       свадьба                        0.080034
2   образование                        0.092200
0    автомобиль                        0.093590

Сводная таблица кредитов с задолженностью в зависимости от цели кредита и уровня дохода клиентов:

In [69]:
pd.pivot_table(df_result, values='debt', index=['purpose_category'], columns=['income_level'], aggfunc={'debt':[np.average]}) \
                  .reset_index() \
                  .rename(columns={"purpose_category": "Цель кредита", "average": "Уровень дохода"})


Цель кредита Уровень дохода                                 
income_level                      высокий    низкий сверхвысокий   средний
0               автомобиль       0.088183  0.084928     0.047619  0.098617
1             недвижимость       0.066901  0.070968     0.047210  0.074699
2              образование       0.083333  0.095355     0.082192  0.093201
3                  свадьба       0.050505  0.087794     0.075000  0.083553

Можно отметить, что клиенты со сверхвысокими доходами чаще имеют задолженности по кредитам в категориях "образование" и "свадьба" (видимо, свадьбы у них - действительно затратное мероприятие :) ). Наименьший процент задолженности свадебных кредитов - у клиентов с высокими доходами (они, видимо, выбирают разумный компромисс в плане стоимости/контента свадьбы)
Процент задолженности в кредитах на образование в целом достаточно высок по всем категориям заемщиков.
Хуже всего с задолженностями по кредиту дело обстоит в категории "автокредит" для клиентов со средним уровнем дохода.

Сводная таблица кредитов с задолженностью в зависимости от цели кредита и наличия детей у клиента:

In [70]:
pd.pivot_table(df_result, values='debt', index=['purpose_category'], columns=['has_children'], aggfunc={'debt':[np.average]}) \
                  .reset_index() \
                  .rename(columns={"purpose_category": "Цель кредита", "average": "Уровень дохода"})

Цель кредита Уровень дохода                      
has_children                  1-2 ребенка многодетные нет детей
0               автомобиль       0.110871    0.083333  0.085413
1             недвижимость       0.081875    0.082051  0.067289
2              образование       0.105873    0.051948  0.086677
3                  свадьба       0.087071    0.138889  0.075163

Можно отметить, что наименьший процент задолженностей по кредитам - для категории "образование" многодетных клиентов (вероятно, данная категория клиентов ответственно подходит к выбору указанного кредитного продукта для своих детей). Хуже всего с задолженностями по кредиту - тоже у многодетных клиентов в категории "свадьба".
В категории "недвижимость" наилучший процент задолженности и клиентов без детей (что тоже вполне объяснимо, поскольку указанной категории клиентов проще рассчитывать свои траты - тем более что в настоящее время многие стараются обзавестись жильем до рождения детей и вероятно, более ответственно относятся к его выплате).
Достаточно удивительно то, что в категории "автокредит" наиболее дисциплинированные клиенты - без детей и многодетные.

## Общий вывод

Для решения задачи и ответов на вопросы исследования представленные данные о клиентах кредитных продуктов банка были предварительно подготовлены и проанализированы. 

На этапе предобработки были устранены пропуски в значениях общего стажа и уровеня ежемесячного дохода. Значения уровня дохода заполнены на основе медианных значений дохода для групп по категориям пол и тип занятости. Также были устранены артефакты и дубликаты в записях, выполнено приведение типов для удобства выполнения последующего анализа.

На этапе категоризации данных с целью ответа на вопросы исследования были определены и заполнены значениями следующие категории данных:
- по количеству детей: нет детей / 1-2 ребенка в семье / многодетные
- по семейному положению (в настоящий момент времени): семейные / холостые
- по семейному положению (когда-либо): состояли (состоят) в браке / не состояли в браке 
- по уровню дохода: низкий / средний / высокий / сверхвысокий
- по целям кредита: недвижимость / автомобиль / образование / свадьба

С целью последующего построения модели кредитного скоринга определены зависимости между наличием задолженностей по кредиту и выделенными категориями данных. Получены следующие результаты:
- по количеству детей: клиенты с 1-2 детьми чаще имеют задолженности по кредитам относительно клиентов без детей и многодетных клиентов
- по семейному положению (в настоящий момент времени): холостые клиенты чаще имеют задолженности по кредитам относительно семейных клиентов
- по семейному положению (когда-либо): клиенты, когда-либо состоявшие в официальном браке, 
- по уровню дохода: клиенты со сверхвысоким уровнем дохода имеют наименьшую задолженность по кредитам, дале следуют клиенты с высоким уровнем дохода. Доля клиентов с задолженностями в сегментах низкого и среднего уровня дохода практически совпадает и выше, чем у клиентов с высокими доходами.
- по целям кредита: наименьший процент задолженности в категориях кредитов на недвижимость, наибольшая - в категориях "автокредит" и "образование".

Также приведены сводные таблицы по проценту задолженности для различных категорий кредита в зависимости от уровня дохода и наличия детей у клиентов. 

По указанной выборке определено, что наименьший процент задолженности у клиентов со сверхвысокими доходами на кредиты в категориях "автомобиль" и "недвижимость", у клиентов с высокими доходами в категории "свадьба", а также у многодетных клиентов в категории "образование".
    
"Худшими" с точки зрения наличия задолженности оказались кредиты, выданные многодетным клиентам на свадьбу, клиентам с 1-2 детьми на автомобиль и образование, а также клиентам со средним уровнем дохода в категории "автомобиль".